letsencrypt/acme/jose/jwa.py

```diff
@@ -0,0 +1,125 @@
+"""JSON Web Algorithm.
+
+https://tools.ietf.org/html/draft-ietf-jose-json-web-algorithms-40
+
+"""
+import abc
+
+from Crypto.Hash import HMAC
+from Crypto.Hash import SHA256
+from Crypto.Hash import SHA384
+from Crypto.Hash import SHA512
+
+from Crypto.Signature import PKCS1_PSS
+from Crypto.Signature import PKCS1_v1_5
+
+from letsencrypt.acme.jose import errors
+from letsencrypt.acme.jose import interfaces
+from letsencrypt.acme.jose import jwk
+
+
+class JWA(interfaces.JSONDeSerializable):  # pylint: disable=abstract-method,too-few-public-methods
+    """JSON Web Algorithm."""
+
+
+class JWASignature(JWA):
+    """JSON Web Signature Algorithm."""
+    SIGNATURES = {}
+
+    def __init__(self, name):
+        self.name = name
+
+    def __eq__(self, other):
+        return isinstance(other, JWASignature) and self.name == other.name
+
+    @classmethod
+    def register(cls, signature_cls):
+        """Register class for JSON deserialization."""
+        cls.SIGNATURES[signature_cls.name] = signature_cls
+        return signature_cls
+
+    def to_json(self):
+        return self.name
+
+    @classmethod
+    def from_json(cls, jobj):
+        return cls.SIGNATURES[jobj]
+
+    @abc.abstractmethod
+    def sign(self, key, msg):  # pragma: no cover
+        """Sign the ``msg`` using ``key``."""
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def verify(self, key, msg, sig):  # pragma: no cover
+        """Verify the ``msg` and ``sig`` using ``key``."""
+        raise NotImplementedError()
+
+    def __repr__(self):
+        return self.name
+
+
+class _JWAHS(JWASignature):
+
+    kty = jwk.JWKOct
+
+    def __init__(self, name, digestmod):
+        super(_JWAHS, self).__init__(name)
+        self.digestmod = digestmod
+
+    def sign(self, key, msg):
+        return HMAC.new(key, msg, self.digestmod).digest()
+
+    def verify(self, key, msg, sig):
+        # TODO: use constant compare to mitigate timing attack?
```

Comment:
This is more important for server side code... however, I am afraid that this library code might be adopted somewhere, given that a good python implementation doesn't already exist.  I would personally make it a warning.
http://www.levigross.com/2014/02/07/constant-time-comparison-functions-in-python-haskell-clojure-java-etc/